### PR TITLE
New completion

### LINKED ISSUE #517: Rewrite CodeCompletion to use new YAML Parser AST
### Is your enhancement related to a problem? Please describe.
With https://github.com/redhat-developer/yaml-language-server/pull/442 we face problems with current code completion implementation. Current implementation doesn’t work with new AST properly.

### Describe the solution you would like
Write new codecompletion implementation which will use AST from new parser.

### PR DESCRIPTION
**Pull request by evidolob**
### What does this PR do?
This PR add usage of new AST for code completion, now code completion uses raw AST(without converting it to JSON compatible AST)  
Basically it is not completely rewrite, it is a port of old code completion to new AST. 
Also to be able select proper JSON Schema I port validation code to new AST(`json-schema07-validator.ts`).
So we can remove AST convert completely( delete `ast-converter.ts` and all related code/types).

### What issues does this PR fix or reference?
Resolve #517

### Is it tested? How?
With tests

### CODE REVIEW

**Review by gorkem — CHANGES_REQUESTED** [review]
Something is not working right with this version. When I use this version of the LS on vscode-yaml, the errors are not cleaned after I fix them until I reopen the editor. 

Also it is it is very hard to identify what to review on this PR because it is mixed with some of the main changes that are coming from main.  Can we seperate the main merge ?

**Comment by evidolob** [comment]
@evidolob I fix validation, now it should work.

File: src/languageservice/services/yamlValidation.ts [hunk 1]
```diff
@@ -62,32 +62,36 @@ export class YAMLValidation {
...
+        index++;
       }
-      if (syd.warnings.length > 0) {
-        validationResult.push(...syd.warnings);
-      }
-
-      validationResult.push(...validation);
-      index++;
+    } catch (err) {
+      console.error(err.toString());
```
Comment (gorkem) [inline src/languageservice/services/yamlValidation.ts:94]: Are we supposed to propogate validation errors to telemetry?

File: src/languageservice/parser/yaml-documents.ts [hunk 2]
```diff
@@ -72,6 +77,103 @@ export class SingleYAMLDocument extends JSONDocument {
...
+        closestNode = node;
+      } else {
+        return visit.SKIP;
+      }
+    });
+
+    return closestNode;
+  }
+
+  findClosestNode(offset: number, textBuffer: TextBuffer): Node {
```
Comment (gorkem) [inline src/languageservice/parser/yaml-documents.ts:102]: Can we add tests for this new function

File: src/languageservice/parser/yaml-documents.ts [hunk 3]
```diff
@@ -72,6 +77,103 @@ export class SingleYAMLDocument extends JSONDocument {
     doc.validate(schema, matchingSchemas, node.start);
     return matchingSchemas;
   }
+
+  getNodeFromPosition(positionOffset: number): Node | undefined {
```
Comment (gorkem) [inline src/languageservice/parser/yaml-documents.ts:81]: Does this new function need tests?

File: src/languageservice/utils/astUtils.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Red Hat, Inc. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Document, isDocument, Node, Pair, visit } from 'yaml';
+
+export function getParent(doc: Document, nodeToFind: Node): Node | undefined {
```
Comment (gorkem) [inline src/languageservice/utils/astUtils.ts:8]: Should be covered with tests

File: src/languageservice/utils/astUtils.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  });
+
+  if (isDocument(parentNode)) {
+    return undefined;
+  }
+
+  return parentNode;
+}
+
+export function getNodePath(doc: Document, node: Node): readonly (Document | Node | Pair)[] | undefined {
```
Comment (gorkem) [inline src/languageservice/utils/astUtils.ts:24]: Should be covered with tests

File: src/languageservice/utils/strings.ts [hunk 6]
```diff
@@ -48,3 +50,18 @@ function convertRegexString2RegExp(pattern: string, flag: string): RegExp {
...
+  }
+
+  for (let i = 0; i < position; i++) {
+    const char = lineContent.charCodeAt(i);
+    if (char !== CharCode.Space && char !== CharCode.Tab) {
+      return i;
+    }
+  }
+
+  return position;
```
Comment (gorkem) [inline src/languageservice/utils/strings.ts:66]: I am not sure why this returns the position here

File: src/languageservice/yamlLanguageService.ts [hunk 7]
```diff
@@ -155,7 +156,8 @@ export function getLanguageService(
   clientCapabilities?: ClientCapabilities
 ): LanguageService {
   const schemaService = new YAMLSchemaService(schemaRequestService, workspaceContext);
-  const completer = new YAMLCompletion(schemaService, clientCapabilities, telemetry);
+  // const completer = new YAMLCompletion(schemaService, clientCapabilities, telemetry);
```
Comment (gorkem) [inline src/languageservice/yamlLanguageService.ts:159]: remove unused code

**Review by gorkem — CHANGES_REQUESTED** [review]
I should still do a bit more testing but functionality seems to be in place. It would be great if we can add tests to the new functions introduced. Otherwise I have some minor comments

File: src/languageservice/parser/yaml-documents.ts [hunk 8]
```diff
@@ -72,6 +77,103 @@ export class SingleYAMLDocument extends JSONDocument {
...
+    let closestNode: Node;
+    visit(this.internalDocument, (key, node: Node) => {
+      if (!node) {
+        return;
```
[the remark this answers is not in the document]
Comment (evidolob) [reply src/languageservice/parser/yaml-documents.ts:85]: No, as according to dock `visit.BREAK`: Terminate traversal completely. 
In this case we return undefined which means `Do nothing and continue`

File: src/languageservice/parser/yaml-documents.ts (hunk 2 above)
Comment (evidolob) [reply src/languageservice/parser/yaml-documents.ts:102]: Done

File: src/languageservice/parser/yaml-documents.ts (hunk 3 above)
Comment (evidolob) [reply src/languageservice/parser/yaml-documents.ts:81]: Yes, I add test for it

File: src/languageservice/utils/astUtils.ts (hunk 4 above)
Comment (evidolob) [reply src/languageservice/utils/astUtils.ts:8]: Done

**Review by gorkem — APPROVED** [review]

**Pull request merged**